app/models/course/material/folder.rb

```diff
@@ -35,4 +55,14 @@ def validate_name_is_unique_among_materials
     conflicts = parent.materials.where('lower(name) = ?', name.downcase)
     errors.add(:name, :taken) if conflicts.any?
   end
+
+  def existing_names
+    parent.children.where.not(id: self).pluck(:name) + parent.materials.pluck(:name)
+  end
+
+  def assign_valid_name
+    return unless name_changed? && owner
```

Comment:
I think it's necessary ? it's only saved once.

Suppose `owner` changes its name to `"AA"`, then folder name will also be changed to `"AA"` in owner's callback. But if `"AA"` is already taken and the folder is not valid, then folder cannot be saved. 

This function helps to update `"AA"` to `"AA(0)"` in this case.